docs/proposals/caching.adoc

```diff
@@ -0,0 +1,77 @@
+= Proposal: Caching
+
+== Purpose
+
+This proposal aims to support caching of dependencies
+(https://github.com/opendevstack/ods-pipeline/issues/147) in order to speed up build times. 
+
+== Background
+
+Currently there is a single PVC per project which is shared by all build pipelines. 
+As a consequence only a single build at a time can run for a project.
+
+The PVC is mounted as the workspace in all build tasks. 
+
+The `ods-start` task wipes the PVC at the beginning of each build so that no data persists between build.
+
+Each tekton task which gets started mounts the workspace PVC which can take quiet noticeable wait times.   
+
+== Solution
+
+In the initial implementation instead of a single PVC per project one PVC per repo will be used. The PVC will be used both as a workspace for the build and also as a cache of data between builds.
+
+A build task run will have the following directories available:
+
+* A repo checkout dir. This is the current workspace directory although I propose to rename it to checkout directory as for some technologies the actual build will not happen here but in a caching directory if available.
+
+With sufficient caching space available: 
+
+* A global cache directory where the parent directory is indicated by parameter `global-cache-parent`. Cleanup will only spare directories with in it so that build tasks must keep stuff in a subdirectory. For example in `$(global-cache-parent)/go-modules/`. 
+
+* A pipeline cache directory as indicated by parameter `pipeline-cache-dir`.
+
+In the initial implementation the parameters will be set to `/.cache/` and `/.cache-p/<pipeline-name>/` if space permits and be empty if there is not sufficient space available.
```

Comment:
Why `/.cache-p/<pipeline-name>/`? Does that mean you want caching within a branch to be on by default?

Reply:
pipeline caching should only be enabled when the tasks declares how much free space it requires for a new pipeline with a parameter `pipeline-caching-mb` greater than 0.
There should probably also be way to not using caching for a specific commit, perhaps by having `[build-no-cache]` as part of a  commit message.
The pipeline cache should be left undisturbed from the previous build of the same pipeline, except in case where cache space is reclaimed. Not sure it answers your first question.